### ISSUE TITLE
bug: virtual networks can't scale if using virtual_network_resource_ids output

### ISSUE THREAD

**Issue by b0bu** [Azure/terraform-azurerm-lz-vending#290]
### Community Note

- Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
- Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
- If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Versions

Please paste the output of `terraform version` command from within the initialized directory:

```text
1.6.4
```

Please enter the module version that you are using:

```text
3.3.0
```

### Description

The output variable https://github.com/Azure/terraform-azurerm-lz-vending/blob/f5d2209403f36186bbec6f0f3f56d26c15a06236/modules/virtualnetwork/outputs.tf#L1 doesn't scale. Maps have no order so add an entry at the beginning middle or end when referencing by key doesn't cause an ordering issue for example 

```hcl
locals {
 mymap = {
 a = 1
 b = 2
 c = 3
 }
}
```
when adding `d = 4` at any position in the map, beginning middle or end doesn't cause references to any existing keys to change based on index position so referencing `local.mymap.a` doesn't change when adding `local.mymap.d`. This doesn't appear to be true for module outputs. The module output  `virtual_network_resource_ids` is output according to alphabetical order. Based on the input `var.virtual_network` if I define an abitrarily ordered map of virtual networks

```hcl
virtual_networks = {
b = {}
c = {} 
a = {}
}
```
the output for `module.sub.virtual_network_resource_ids` will always be 
```hcl
output vnets {
 value = module.sub.virtual_network_resource_ids
}
vnets = {
  a = id
  b = id
  c = id
}
```

This means that if you're using the outputs from the module to create subnets, and you scale the virtual network object by adding a vnet at any position, the order of all virtual networks is updated or becomes unknown to terraform. So if I'm referencing the vnet id via the module output and I add a vnet, anything using that references loses it resulting in a `(known after apply) # force replacement` in the plan. 

#### Steps to Reproduce

1. create some virtual networks 
```hcl
module sub {
  ...
  virtual_networks = {
    vnet1 = {
     name = "vnet1"
     address_space = ["10.10.0.0/16]
     location = "uksouth"
     resource_group_name = module.rg.name
     ...
    }
    vnet2 = {
     name = "vnet2"
     address_space = ["10.20.0.0/16"]
     location = "uksouth"
     resource_group_name = module.rg.name
     ...
    }
  }
 ...
}
```
2. use the module output
```hcl
locals {
  private_zones = {
    vault-private = {
      zone = "privatelink.vaultcore.azure.net"
    }
    vault-public = {
      zone = "vaultcore.azure.net"
    }
  }
}

resource azurerm_private_dns_zone_virtual_network_link "link" {
  for_each = local.private_zones
  name = each.key
  resource_group_name = module.rg.name
  private_dns_zone_name = each.value.zone
  virtual_network_id = module.sub.virtual_network_resource_ids.vnet1
}
```
3. update the virtual networks
```hcl
module sub {
  ...
  virtual_networks = {
    vnet1 = {
     name = "vnet1"
     address_space = ["10.10.0.0/16"]
     location = "uksouth"
     resource_group_name = module.rg.name
     ...
    }
    vnet2 = {
     name = "vnet2"
     address_space = ["10.20.0.0/16"]
     location = "uksouth"
     resource_group_name = module.rg.name
     ...
    }
    vnet3 = {
     name = "vnet3"
     address_space = ["10.30.0.0/16"]
     location = "uksouth"
     resource_group_name = module.rg.name
     ...
    }
  }
 ...
}
```
4. reference to `module.sub.virtual_network_resource_ids.vnet1` is now lost and forces replacement of all dns links

#### Additional context
The only "fix" I've found for this is to create a copy of the output as a local
```hcl
locals {
  vnets = {
    vnet1 = id
    vnet2 = id 
  ...
  }
}
```
that way adding a vnet3 doesn't lose the reference to vnet1 as it's a static map and not dynamically and without any imposed order but this obviously breaks the dependency tree and updating things becomes painful

**Comment by matt-FFFFFF** [comment]
Hi @b0bu - this is an interesting find, thanks for pointing it out.

Would it help if we created an unordered output, e.g. set of objects? We might have the same issue in that the entire set may be `known after apply`

**Comment by b0bu** [comment]
@matt-FFFFFF I pulled the code this afternoon and had a bit of a play, even outputting a tuple of objects preserves the alphabetical order which was quite frustrating. I'm wondering if the ordering problem is happening at azapi_resource.vnet create and might need to change how those resources are being created i.e. `for_each = var.virtual_networks` as a data structure. I'll do some digging on it but I suspect the issue is terraform internal structs and converting data types. This has to be a solved problem though.

**Comment by matt-FFFFFF** [comment]
I have tried to repo using a child module and a parent module. The child module created terraform_data resources and a local map. The parent module creates terraform_data resources based off the output of the child module (a map).

I do not get the ordering problem with that basic config.

More work is needed to understand why the output is marked known after apply.

**Comment by b0bu** [comment]
I did an experiment where I changed the map of objects to a list of objects in the hopes that adding `vnet_uks_3` would stop the output from being destroyed.

```hcl
  virtual_networks = [
    {
      name          = "vnet_uks_1"
      address_space = ["10.10.0.0/24"]
      location      = "uksouth"

      vwan_connection_enabled = false

      resource_group_name             = module.sandbox_vnets_uks_rg.name
      tags                            = local.default_tags
      resource_group_creation_enabled = false
    },
    {
      name          = "vnet_uks_2"
      address_space = ["10.20.0.0/24"]
      location      = "uksouth"

      vwan_connection_enabled = false

      resource_group_name             = module.sandbox_vnets_uks_rg.name
      tags                            = local.default_tags
      resource_group_creation_enabled = false
    }
  ]
```
the output now looks like this so that the vnet can be referenced by its name at least by `v.name`
```hcl
output "virtual_network_resource_ids" {
  description = "The created virtual network resource IDs, expressed as a map."
  value = {
    for k, v in azapi_resource.vnet : v.name => v.id 
  }
}
```
like this 
```hcl
resource "azurerm_private_dns_zone_virtual_network_link" "link" {
  for_each              = azurerm_private_dns_zone.zones
  name                  = each.key
  resource_group_name   = module.sandbox_vnets_uks_rg.name
  private_dns_zone_name = trimsuffix(each.value.soa_record[0].fqdn, ".")
  virtual_network_id    = module.sub.virtual_network_resource_ids.vnet_uks_1
  tags                  = local.default_tags
}
```
however the plan still resulted in the exact issue with `2 to destroy.` being the network links for the private zones. And as you can see, the output object. 
```
Plan: 4 to add, 0 to change, 2 to destroy.

Changes to Outputs:
  ~ vents                      = {
      - vnet_uks_1 = "/subscriptions/000/resourceGroups/sanbox-vnets-uks/providers/Microsoft.Network/virtualNetworks/vnet_uks_1"
      - vnet_uks_2 = "/subscriptions/000/resourceGroups/sanbox-vnets-uks/providers/Microsoft.Network/virtualNetworks/vnet_uks_2"
    } -> (known after apply)
```
this is the exact same result if I leave them index based
```
Plan: 4 to add, 0 to change, 2 to destroy.

Changes to Outputs:
  ~ vents                      = {
      - "0" = "/subscriptions/000/resourceGroups/sanbox-vnets-uks/providers/Microsoft.Network/virtualNetworks/vnet_uks_1"
      - "1" = "/subscriptions/000/resourceGroups/sanbox-vnets-uks/providers/Microsoft.Network/virtualNetworks/vnet_uks_2"
    } -> (known after apply)
```
although interestingly, if I just let this create destroying my network links (fine in testing when I have 2 but in reality I have about 500 in production) so now I have the 3 networks in place 

commenting out the vnet_uks_3 works as expected, it removes the map entry from the set, leaving 2 unchanged attributes and the 2 resources destroyed are `azapi_resource.vnet` and `azapi_update_resource.vnet` as expected. 
```
Plan: 0 to add, 0 to change, 2 to destroy.

Changes to Outputs:
  ~ vents                      = {
      - "vnet_uks_3" = "/subscriptions/000/resourceGroups/sanbox-vnets-uks/providers/Microsoft.Network/virtualNetworks/vnet_uks_3"
        # (2 unchanged attributes hidden)
    }
```
why this doesn't work the same when adding the network I still do not know, but a list of objects seems to produce the same result...

**Comment by b0bu** [comment]
@matt-FFFFFF https://github.com/Azure/terraform-azurerm-lz-vending/blob/67c456b468c20f6bd9a7b2fa5659c4af9593bec4/locals.tf#L27 this is the issue. 

changing 
```hcl
  virtual_network_resource_ids = try(module.virtualnetwork[0].virtual_network_resource_ids, {})
```
to 
```hcl
  virtual_network_resource_ids = module.virtualnetwork[0].virtual_network_resource_ids
```
fixes the issue. Adding to `var.virtual_networks` causes try to fall back to `{}` destroying the output. I haven't put a PR to resolve since I'm not certain what the intention with `try()` was. It looks like you're saying that you always want `virtual_network_resource_ids` to have a value even if it's empty.

**Comment by matt-FFFFFF** [comment]
This is excellent work @b0bu - thank you

I suspect the `try()` was put in to cater for the scenario where the vnet submodule was not used.

If we remove the try we need to ensure that we don't get an error if we are not deploying vnets to the sub. I can test this over the next week.

**Comment by b0bu** [comment]
Yup makes sense. Having `locals` as an intermediary layer between parent and submodule is a clever way to conditionally toggled features. 

changing this from
```hcl
 virtual_network_resource_ids = try(module.virtualnetwork[0].virtual_network_resource_ids, {}) 
```
to
```hcl
 virtual_network_resource_ids = var.virtual_network_enabled ? module.virtualnetwork[0].virtual_network_resource_ids : {}
```
fixes the issue

The `locals` layer might still be a valid and useful pattern for some things but I wonder if this change makes its abstraction needlessly complex.

**Comment by b0bu** [comment]
@matt-FFFFFF is this something you want me to go ahead with or are you rolling it up into other changes?

**Comment by matt-FFFFFF** [comment]
Hi @b0bu 

I'll get around to this by the end of the week. Apologies for the silence there has been some annual leave in the team.

**Comment by b0bu** [comment]
@matt-FFFFFF any update? Sorry for being a pain

**Comment by matt-FFFFFF** [comment]
Just released as v3.4.2
